Serial driver: major rewrite

leamas opened the pull request:
Major changes to the serial driver. Here is:
  - Use faster and simpler line-oriented buffer instead of character based ones. 
  - Refactor the IO thread to a separate file.
  - Avoid the new/delete allocation of the thread, use an automatic stack variable instead.
  - Use a new ThreadCtrl mixin class to avoid the slow and somewhat clumsy polling when closing IO thread
  - When receiving messages, avoid the extra message passing  and thus the extra event loop  cycle in current code.
  - Enhance sanity checks to avoid passing invalid data around.

Closes: #3116 
Closes: #3866

leamas commented:
rebased: fix windows builds.

leamas commented:
Tested on windows and Linux, working

leamas commented:
found some glitches...

bdbcat commented:
Builds but crashes at runtime on Android.
`java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_ZN27CommDriverN0183SerialThreadC1ENSt6__ndk18functionIFvRKNS0_6vectorIhNS0_9allocatorIhEEEEEEE"`

Probably causes by initialization of member in driver ctor, but the SerialThread objects are #ifdef'ed away for Android? 
```

CommDriverN0183Serial::CommDriverN0183Serial(const ConnectionParams* params,
                                             DriverListener& listener)
    : CommDriverN0183(NavAddr::Bus::N0183,
                      ((ConnectionParams*)params)->GetStrippedDSPort()),
      m_portstring(params->GetDSPort()),
      m_secondary_thread(
          [&](const std::vector<unsigned char>& v) { SendMessage(v); }),


```
Sorry

bdbcat commented:
Some experimentation.
After fussing around to remove  CommDriverN0183SerialThread, with some things like:

```
#ifndef __ANDROID__
  CommDriverN0183SerialThread m_secondary_thread;
#endif

```
I did succeed in building and running.  Works fine, as expected.

leamas commented:
Thanks!

This then boils down to make it compile and run on Android in a reasonably clean way. Could you perhaps share your diff?

I have some ideas how to handle this, but #4164 takes some toll. Real  life also needs some attention.

bdbcat commented:
Here is my diff.
Does not build on linux, not sure how to best handle the initialization of serial thread on driver CTOR.
But it does load and run correctly on Android.
Hope it helps.

```
diff --git a/model/include/model/comm_drv_n0183_serial.h b/model/include/model/comm_drv_n0183_serial.h
index 791cb4d96..a435016b1 100644
--- a/model/include/model/comm_drv_n0183_serial.h
+++ b/model/include/model/comm_drv_n0183_serial.h
@@ -46,8 +46,9 @@ public:
   bool Open();
   void Close();
 
+#ifndef __ANDROID__
   bool IsSecThreadActive() { return m_secondary_thread.IsRunning(); }
-
+#endif
   bool IsGarminThreadActive();
   void StopGarminUSBIOThread(bool bPause);
 
@@ -66,7 +67,9 @@ private:
   std::string m_portstring;
   std::string m_baudrate;
 
-  CommDriverN0183SerialThread m_secondary_thread;
+#ifndef __ANDROID__
+    CommDriverN0183SerialThread m_secondary_thread;
+#endif
   GarminProtocolHandler* m_garmin_handler;
 
   ConnectionParams m_params;
diff --git a/model/src/comm_drv_n0183_serial.cpp b/model/src/comm_drv_n0183_serial.cpp
index b8ddacf01..80b4966ad 100644
--- a/model/src/comm_drv_n0183_serial.cpp
+++ b/model/src/comm_drv_n0183_serial.cpp
@@ -66,8 +66,8 @@ CommDriverN0183Serial::CommDriverN0183Serial(const ConnectionParams* params,
     : CommDriverN0183(NavAddr::Bus::N0183,
                       ((ConnectionParams*)params)->GetStrippedDSPort()),
       m_portstring(params->GetDSPort()),
-      m_secondary_thread(
-          [&](const std::vector<unsigned char>& v) { SendMessage(v); }),
+ //     m_secondary_thread(
+ //         [&](const std::vector<unsigned char>& v) { SendMessage(v); }),
       m_params(*params),
       m_listener(listener) {
   m_baudrate = wxString::Format("%i", params->Baudrate);

```

leamas commented:
I have Ideas, but no time. Stay tuned.

bdbcat commented:
Now I am thinking that there is no problem to define and instantiate the secondary thread object for Android.
Just don't start it, ever, for Android.
Might save some evil ifdefs...

leamas commented:
This is more or less what I'm doing right now...

leamas commented:
This is what I have right now. Not tested at all, will have opportunities to test later today. That is, still a sketch.

The former "thread" class is now just an io handler. For non-android platform it runs a thread, but android does not. They both have the same interface.

Almost all #ifdefs gone, one annoying one left which I'm thinking about.

bdbcat commented:
Sorry, no go.
UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_ZTI8SerialIo" referenced by "/data/app

In Android build, in the CTOR of serial driver, I do not see how it can find  an abstract Serlalio() to initialize the member variable as:
      m_serial_io(SerialIo::Create(

Or something like that...

bdbcat commented:
Sorry, same result.
It must be due to clang vs gcc.
Side:  It is no trouble to test patches like this.  Takes me about 5 minutes to rebuild and smoke test.  Bring 'em on.

leamas commented:
Digging... I can see the symbol in the opencpn executable:  `objdump -T -C | grep SerialIo`  contains this line
```
0000000000ca0e80  w   DO .data.rel.ro   0000000000000038  Base   vtable for SerialIo
```
And yes, running without -C shows that `vtable for SerialIo`is the same thing  (same address) as _ZTV8SerialIo.

That is, `vtable for SerialIO` a k a  _ZTV8SerialIo  is defined but weak.  Trying to read about Android dl-loading of weak symbols, here seems to be dragons.  Current state: I have read some links, but honestly don't understand. 

Does this ring any bells for you?

bdbcat commented:
For another look:

```
dsr@debian:~/Projects/opencpn/build_android_32_cci_22b$ readelf -s libgorp.so | grep SerialIo
   387: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND _ZTV8SerialIo
   436: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND _ZTI8SerialIo
  6689: 00dae2b4   128 FUNC    WEAK   DEFAULT   14 _ZN8SerialIoC2Ev
 16366: 017fdaf0    28 OBJECT  WEAK   DEFAULT   16 _ZTV15AndroidSerialIo
 25772: 0071fa03    18 OBJECT  WEAK   DEFAULT   13 _ZTS15AndroidSerialIo
 31554: 00ed8fb0   180 FUNC    WEAK   DEFAULT   14 _ZN8SerialIoC2EN[...]
 40414: 017fdb0c    12 OBJECT  WEAK   DEFAULT   16 _ZTI15AndroidSerialIo
 48821: 00ed8df8   232 FUNC    GLOBAL DEFAULT   14 _ZN8SerialIo6Cre[...]
 49914: 00ed9448   112 FUNC    WEAK   DEFAULT   14 _ZN8SerialIoD2Ev
207934: 00ed8df8   232 FUNC    GLOBAL DEFAULT   14 _ZN8SerialIo6Cre[...]
207935: 00dae2b4   128 FUNC    WEAK   DEFAULT   14 _ZN8SerialIoC2Ev
207949: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND _ZTV8SerialIo
211271: 00ed8fb0   180 FUNC    WEAK   DEFAULT   14 _ZN8SerialIoC2EN[...]
211272: 00ed9448   112 FUNC    WEAK   DEFAULT   14 _ZN8SerialIoD2Ev
211278: 017fdb0c    12 OBJECT  WEAK   DEFAULT   16 _ZTI15AndroidSerialIo
211279: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND _ZTI8SerialIo
211280: 0071fa03    18 OBJECT  WEAK   DEFAULT   13 _ZTS15AndroidSerialIo
211281: 017fdaf0    28 OBJECT  WEAK   DEFAULT   16 _ZTV15AndroidSerialIo
```
Maybe nothing to do with weak symbol, just cannot find SerialIo object at all.
I tried adding a default constructor for SerialIo.  No help.
Anyway, with this readelf output, it is bound to fail on dlopen() of libgorp.so

Maybe time for an evil #ifdef in the driver member definition, moving the item away from abstract SerialIo, down to concrete Android or std definition.  To do that, we need a header file for android_serial_io.cpp.

Head scratch....

leamas commented:
OTOH, after building the "fake" armhf opencpn in a docker container:

```
# readelf -s  opencpn | grep SerialIo | grep UND                                                                          
# 
```

I. e., these symbols are are actually defined when building opencpn, but not when building libgorp.so. Is this about how libgorp .so is linked? 

In particular, what if (shot in the dark):
  1. android_serial_io is first linked resolving  all internal references
  2. comm_drv_n0183_serial is linked after android_serial_io. It will need SerialIo, but since it was not unresolved  when linking  android_serial_io it's not available and becomes UNDEF.
 

I  used to scratch my hair. These days, scratching my head.

leamas commented:
Forget it. I have been building the wrong thing. Stay tuned

leamas commented:
I can reproduce it. It's not a link ordering problem.

leamas commented:
Tracked it down to the android_serial_io library:  `readelf -s -C ./model/CMakeFiles/_model_src.dir/src/android_serial_io.cpp.obj | grep SerialIo` gives 

```
89: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND typeinfo for SerialIo
93: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND vtable for SerialIo
```
 and a bunch of defined ones. When building for Linux, std_serial_io does not exhibit this, here everything is defined.

Now off for other duties.

leamas commented:
Android linkage error seems to be fixed by hack in [1c3eb99](https://github.com/OpenCPN/OpenCPN/pull/4160/commits/1c3eb99302a08edbe447cadf8b4f26a27c8531c5). 

Found another issue, fixed and pushed. Ready for new Android tests, needs to make more windows testing myself

bdbcat commented:
Now getting very close.  Builds, installs, and runs on Android.
One Android problem, in this function:

```
static std::string NormalizePort(const std::string& port_id) {
  std::string port(port_id);
  if (ocpn::startswith(port, "Serial:")) port = ocpn::split(port, ":")[1];
  return port;
}

```
Returns the "wrong" result with input "Serial:AUSBSerial:FTDI_FT232R".  The result required for Android is"
"AUSBSerial:FTDI_FT232R".
Easy enough with wxString().  wxString AfterFirst(':').....

Otherwise, I think ready.

leamas commented:
Or IMHO patch below. The wxWidgets docs actually states that using wxString is only recommended when interfacing wxWidgets itself.

Squashed this into top commit, cleaned up the hack to 3c8217aa7 and rebased to current master

```
diff --git a/model/include/model/serial_io.h b/model/include/model/serial_io.h
index de577f87b..d9c1d001f 100644
--- a/model/include/model/serial_io.h
+++ b/model/include/model/serial_io.h
@@ -35,10 +35,9 @@
 #include "model/thread_ctrl.h"
 #include "model/ocpn_utils.h"
 
-static std::string NormalizePort(const std::string& port_id) {
-  std::string port(port_id);
-  if (ocpn::startswith(port, "Serial:")) port = ocpn::split(port, ":")[1];
-  return port;
+/** Remove possible Serial: prefix. */
+static std::string NormalizePort(const std::string& port) {
+  return port.find("Serial:") == 0 ? port.substr(strlen("Serial:")) : port;
 }
 
 /** Forwards data from driver to local receivers (main code, plugins, etc). */

```

bdbcat commented:
Sure, I agree to stay with std::string.   wxString example just to show directly what the result should be.
Lots of new commits here.  Let me know when to test again before merging....

leamas commented:
Ready to test.  Actually no new commits, just rebased + minor cleanup of the two top commits

leamas commented:
Sidenote: Reviewing the rebase I note that largest part of the diff is due to clang-format being applied to the changes in #4172.  This makes the diff much bigger than it should be and thus hard to read. The root cause is that clang-format wasn't applied to 4172 before merging. We should IMHO make sure this does not happen again. (@Hakansv )

bdbcat commented:
All good now.  Ready to merge, from my Android perspective.

leamas commented:
Looks good here as well. Merging

leamas commented:
Holding merge until master ci builds are restored

Hakansv commented:
I've tested on Win and all works fine. Also Com ports with number >=10 that historical have been tricky. Great!
One comment though.
The new message when a stated com port is for any reason not connected is OK once. 
But the log file flooding after that message may not be necessary?
The below message can be accepted once. But not hundred in minutes? I think O should with great patience and silence wait for the USB to be connected again.
19:26:57.376 MESSAGE std_serial_io.cpp:162 Unhandled Exception while opening serial port: IO Exception: Specified port, COM5, does not exist., file C:\Builds\OCPN\OpenCPN\libs\serial\src\impl\win.cc, line 92.

Hakansv commented:
One may have expected a "normal" OCPN OK-button? Now I have to use the upper right cross to close it, If don't like to wait for the time out, and that cross is so to say not usual to use on O. 
![bild](https://github.com/user-attachments/assets/8307191b-5a60-4d03-964d-ea47cacde41e)

leamas commented:
yes, I have also noted that. Rough edges to fix, seems to be some bug the supposedly "standard" dialog box...

leamas commented:
The log message needs some care, agreed
